CSharpStepTemplate.yml

```diff
@@ -0,0 +1,142 @@
+- checkout: SdkGenerator
+  clean: true
+  path: s/SdkGenerator
+- checkout: CSharpSdk
+  clean: true
+  submodules: true
+  path: s/sdks/CSharpSDK
+  persistCredentials: true
+- bash: |
+    set -e
+    
+    echo alias the ADO variables into local variables
+    ApiSpecSource="${{ parameters.ApiSpecSource }}"
+    CommitMessage="${{ parameters.CommitMessage }}"
+    GitDestBranch="${{ parameters.GitDestBranch }}"
+    SdkName="${{ parameters.SdkName }}"
+    TestCompile="${{ parameters.TestCompile }}"
+    TestWin32Basic="${{ parameters.TestWin32Basic }}"
+    TestWin32NetCore="${{ parameters.TestWin32NetCore }}"
+    WORKSPACE=$(pwd -W)
+    # Hack attempt to get WORKSPACE into a sub-environment
+    export WORKSPACE="$WORKSPACE"
+    
+    echo === load util.sh to find msbuild ===
+    if [ -f "util.sh" ]; then
+      . "./util.sh" 2> /dev/null
+    elif [ ! -z "$WORKSPACE" ]; then
+      . "$WORKSPACE/JenkinsSdkSetupScripts/JenkinsScripts/Pipeline/util.sh" 2> /dev/null
+    fi
+    
+    Find2019MsBuild || Find2017MsBuild
+    
+    cd "$WORKSPACE/sdks/$SdkName"
+    
+    . "$WORKSPACE/JenkinsSdkSetupScripts/JenkinsScripts/Pipeline/testInit.sh"
+    
+    cd "$WORKSPACE/SDKGenerator/SDKBuildScripts"
+    targetSrc="csharp"
+    export PF_TEST_TITLE_DATA_JSON="$WORKSPACE\JenkinsSdkSetupScripts\Creds\testTitleData.json"
+    . ./shared_build.sh
+
+    echo === Build the CSharp Project ===
+    if $TestCompile
+    then
+      "$MSBUILD_EXE" "$WORKSPACE//sdks//$SdkName//PlayFabSDK//PlayFabSDK.sln" //restore //p:configuration="Debug" //t:Rebuild
+    fi
+    
+    cd "$WORKSPACE/sdks/$SdkName/PlayFabSDK"
+    
+    echo === Build the CSharp UnitTestRunner ===
+    if $TestWin32NetCore
+    then
+      "$MSBUILD_EXE" "$WORKSPACE//sdks//$SdkName//PlayFabSDK//UnittestRunner//UnittestRunner.csproj" //restore //p:configuration="Release" //t:Rebuild
+      echo === Run the CSharpSdk UnitTestRunner net core app 2.0 ===
+      cd "$WORKSPACE/sdks/$SdkName/PlayFabSDK/UnittestRunner/bin/Release/netcoreapp2.0"
+      cmd <<< dotnet .\\UnittestRunner.dll -testInputsFile "$WORKSPACE\\JenkinsSdkSetupScripts\\Creds\\testTitleData.json"
+    fi
+
+    echo === Build PlayFabSDK+Unit.sln ===
+    if $TestWin32Basic
+    then
+      "$MSBUILD_EXE" "$WORKSPACE//sdks//$SdkName//PlayFabSDK//PlayFabSDK+Unit.sln" //restore //p:configuration="Release" //t:Rebuild
+    fi
+
+    echo === Run PlayFabSDK+Unit.sln ===
+    cd "$WORKSPACE//sdks//$SdkName//PlayFabSDK//UnittestRunner//bin//Release//net45"
+    cmd <<< UnittestRunner.exe -testInputsFile "$WORKSPACE//JenkinsSdkSetupScripts//Creds//testTitleData.json"
+
+    echo === Nuget Restore JenkinsConsoleUtility ===
+    cd "$WORKSPACE//SDKGenerator//JenkinsConsoleUtility"
+    nuget restore JenkinsConsoleUtility.sln
+
+    echo === Build JenkinsConsoleUtility ===
+    "$MSBUILD_EXE" "$WORKSPACE//SDKGenerator//JenkinsConsoleUtility//JenkinsConsoleUtility.sln" //p:configuration="Debug" //p:platform="Any CPU" //t:Rebuild
```

Comment:
```suggestion
```
ListenCsJCU does this automatically, and only when it's necessary (it will only build once, if it's not already built)